src/lib.rs

```diff
@@ -48,63 +48,90 @@ pub struct rustls_client_session {
 const RUSTLS_CRATE_VERSION: &str = "0.19.0";
 
 #[macro_export]
-macro_rules! ffi_panic_boundary {
-    ( $($tt:tt)* ) => {
+macro_rules! ffi_panic_boundary_generic {
+    ( $retval:expr, $($tt:tt)* ) => {
         match ::std::panic::catch_unwind(|| {
             $($tt)*
         }) {
             Ok(ret) => ret,
-            Err(_) => return rustls_result::Panic,
+            Err(_) => return $retval,
         }
-  }
+    }
 }
 
 #[macro_export]
-macro_rules! ffi_panic_boundary_size_t {
+macro_rules! ffi_panic_boundary {
     ( $($tt:tt)* ) => {
         match ::std::panic::catch_unwind(|| {
             $($tt)*
         }) {
-            | Ok(ret) => ret,
-            | Err(_) => return 0,
+            Ok(ret) => ret,
+            Err(_) => return rustls_result::Panic,
         }
   }
 }
 
+#[macro_export]
+macro_rules! ffi_panic_boundary_size_t {
+    ( $($tt:tt)* ) => {
+        ffi_panic_boundary_generic!(0, $($tt)*)
+    }
+}
+
 #[macro_export]
 macro_rules! ffi_panic_boundary_bool {
     ( $($tt:tt)* ) => {
-        match ::std::panic::catch_unwind(|| {
-            $($tt)*
-        }) {
-            | Ok(ret) => ret,
-            | Err(_) => return false,
-        }
-  }
+        ffi_panic_boundary_generic!(false, $($tt)*)
+    }
 }
 
 #[macro_export]
 macro_rules! ffi_panic_boundary_ptr {
     ( $($tt:tt)* ) => {
-        match ::std::panic::catch_unwind(|| {
-            $($tt)*
-        }) {
-            | Ok(ret) => ret,
-            | Err(_) => return std::ptr::null_mut(),
-        }
-  }
+        ffi_panic_boundary_generic!(std::ptr::null_mut(), $($tt)*)
+    }
 }
 
 #[macro_export]
 macro_rules! ffi_panic_boundary_unit {
     ( $($tt:tt)* ) => {
-        match ::std::panic::catch_unwind(|| {
-            $($tt)*
-        }) {
-            | Ok(ret) => ret,
-            | Err(_) => return,
-        }
-  }
+        ffi_panic_boundary_generic!((), $($tt)*)
+    }
+}
+
+/// If the provided pointer is non-null, convert it to a reference
+/// to the type in the second argument.
+/// Otherwise, return NullParameter (in the two-argument form) or the provided
+/// value (in the three-argument form).
+#[macro_export]
+macro_rules! checked_ptr_ref {
+    ( $var:ident, $typ:ty ) => {
+        checked_ptr_ref!($var, $typ, rustls_result::NullParameter)
+    };
+    ( $var:ident, $typ:ty, $retval:expr ) => {
+        unsafe {
+            match ($var as *mut $typ).as_ref() {
```

Comment:
Does this need to be `as *mut $typ` or would `as *$typ` suffice? Put otherwise: since we're ultimately getting an immutable reference, why do we need a mutable reference to `var`?

Reply:
Good catch. This actually shouldbe `*const $typ`.